Add positions system to keep track of object' locations

GiudGiud opened the pull request:
This adds a couple of the ones I could think of during the design phase. 

~~We should probably wait to deploy this until it can be used to create multiapps~~

~~Still missing docs~~

This picture explains a bit what is happening with positons in transfers. 
The nearest node line would be in the middle of the two subapps
But the positions are on the bottom left of each app, so the "nearest position" is used to preselect where the points come from

![save](https://user-images.githubusercontent.com/25037151/235244608-90e6410a-a54a-48c0-be3b-92e55b818b57.png)


The best picture for this would involve fuel pins, eg deployment to reactor cases. So it wont be there for a bit

Refs
https://github.com/idaholab/moose/issues/23587

maxnezdyur committed:
create specializations for LineSegment (#23736)

maxnezdyur committed:
Apply suggestions from code review.

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

maxnezdyur committed:
add point to constant reporter

GiudGiud committed:
Add Position creation action, refs #23587

GiudGiud committed:
Add base Position class
- initial position
- multi D options w/ unrolling utility

loganharbour committed:
Finalize datastore/load and to_json for LineSegment and Point

GiudGiud committed:
Add input position test

GiudGiud committed:
Add positions from input parameter

GiudGiud committed:
Add test for positions from file

GiudGiud committed:
Tests file with all test specs

GiudGiud committed:
Add Positions from file

GiudGiud committed:
Remove param group that is not helping for MGs
- it s hiding the dim parameter at the bottom of the list
- it s just one param, no need for a group
Typo fixes

GiudGiud committed:
Add positions from MultiApps, refs #23587

GiudGiud committed:
Add positions from Reporters, refs #23587

GiudGiud committed:
Add centroid positions: right now just elements

GiudGiud committed:
Add test for positions from element centroids

GiudGiud committed:
Add test for positions of groups of elements, refs #23587

GiudGiud committed:
Add positions from groups of elements

GiudGiud committed:
Add test for positions from multiapps

GiudGiud committed:
Add class descriptions to Positions objects

GiudGiud committed:
Use execution as a way to refresh the positioons

GiudGiud committed:
Add capability to generate multiapps from Positions objects

GiudGiud committed:
Add base documentation for the Positions system

GiudGiud committed:
Add object documentation for Positions objects

GiudGiud committed:
Add test for reporter positions

GiudGiud committed:
Deprecate old positions parameter so we can clean up the code later

GiudGiud committed:
Reset positions vectors before filling them using push_back

GiudGiud committed:
Add proper getter for Positions objects
Position -> Positions

GiudGiud committed:
Finalize support for multiapp from positions
- intialize MOST positions as early as possible: at construction

GiudGiud committed:
Make checks on initial positions statuses part of runtime:
users can trigger them with various sitations (notably initial_pos and pos not of same size)

GiudGiud committed:
Add test for creating multiapps from Postions

GiudGiud committed:
Add all gold files for tests

GiudGiud committed:
Create a PositionsName string type to avoid more restrictive reporter syntax

GiudGiud committed:
Sort parameters in MultiApp
Prepare for deprecation of centroid multiapp and Qp multiapp

GiudGiud committed:
Add simple postprocessor to count Positions

GiudGiud committed:
Add ability to sort and broadcast positions

GiudGiud committed:
Add basic routine to find the closest position
We could get fancier with a KDTree approach for example. This will do for now

GiudGiud committed:
Whitespace

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1478871/) on 9f2feac wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24071/site)

This comment will be updated on new commits.

GiudGiud committed:
Add an option to get the multiapp mesh centroid rather than the position

GiudGiud committed:
Avoid adding app centroid positions several times

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 5e720a226ba3740dbe92c135936612461f489ada
+Subproject commit 482ca36f0ba05dbb5d87d02688d43eaa4bf983eb
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1482043/) on 44a9f09 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpich/conda_build_config.yaml
  conda/libmesh/meta.yaml

The libmesh submodule or configuration was changed but the conda build config was not

GiudGiud committed:
Fix documentation failures

GiudGiud commented:
Should be ready for review. Please let me know if you'd rather I break this down in several PRs between:
- adding the positions system
- using it to create multiapps
- using it for nearest node transfers to match origin & target nodes

GiudGiud committed:
Add documentation for the num positions PP

lindsayad commented:
> using it for nearest node transfers to match origin & target nodes

I think I would like to see this split out. I don't currently understand the figure in your title post, but I don't need to for reviewing the generic positions system.

GiudGiud commented:
Sounds good I ll break it off today

friedmud commented:
Excellent work @GiudGiud !  Thanks for prioritizing this... and congrats on the new System ;-)

lindsayad commented on framework/doc/content/source/positions/ElementGroupCentroidPositions.md:
```diff
@@ -0,0 +1,59 @@
+respectively in the Reactor module.
+
+Each block or extra element id specified in the parameters creates new
+bins, it does not indicate a wider group of element to compute the centroid of.
```
```suggestion
bins; it does not indicate a wider group of element to compute the centroid of.
```

lindsayad commented on framework/doc/content/syntax/Positions/index.md:
Cannot currently get to this page from the Sources page. Will need to update the yaml

lindsayad commented on framework/doc/content/source/positions/MultiAppPositions.md:
```diff
@@ -0,0 +1,25 @@
+
+`MultiApps` have a position that is used to translate their frame of reference compared to
+the main app frame of reference. This position can be used to align the meshes for the transfers or
+for outputting the `MultiApps` [Exodus results](Exodus.md) at the translated positions.
```
Does this leverage the coord transform object?

GiudGiud replied:
I ll have to think about it. Depends on the order of operations I guess.

I would say no for the centroid case for sure

lindsayad commented on framework/doc/content/source/postprocessors/NumPositions.md:
```diff
@@ -0,0 +1,13 @@
+
+!alert warning
+Currently, the number of positions should be constant throughout the simulation.
+Please contact a MOOSE developers if you need support for variable numbers of Positions.
```
Like over the course of a transient?

GiudGiud replied:
yeah

lindsayad commented on framework/include/dirackernels/ReporterPointSource.h:
```diff
@@ -37,6 +37,10 @@ class ReporterPointSource : public DiracKernel, public ReporterInterface
   const std::vector<Real> & _y_coord;
   ///z coordinate
   const std::vector<Real> & _z_coord;
+  const std::vector<Point> & _point;
   /// map to associate points with their index into the vpp value
   std::map<Point, size_t> _point_to_index;
+
+  const std::vector<Real> _empty_vec = {};
+  const std::vector<Point> _empty_points = {};
```
doxygen

GiudGiud replied:
this code is from @loganharbour and @maxnezdyur 's commits. I dont really want to modify it so it doesnt create conflicts with their PR (which will get in one day?)

lindsayad replied:
I don't think yours is going to take that long to get in...

GiudGiud replied:
removed them

lindsayad commented on framework/include/multiapps/MultiApp.h:
```diff
@@ -461,8 +467,12 @@ class MultiApp : public MooseObject,
   std::string _app_type;
 
-  /// The positions of all of the apps
+  /// The positions of all of the apps (to be deprecated)
```
```suggestion
  /// The positions of all of the apps, using input constant vectors (to be deprecated)
```

lindsayad commented on framework/include/multiapps/MultiApp.h:
```diff
@@ -461,8 +467,12 @@ class MultiApp : public MooseObject,
+  /// The offsets, in case multiple Positions objects are specified
+  std::vector<unsigned int> _positions_index_offsets;
 
   /// Toggle use of "positions"
```
This doxygen seems ambiguous given the sweeping additions you're making. I might flesh this doxygen comment out a bit

lindsayad commented on framework/include/outputs/JsonIO.h:
```diff
@@ -12,7 +12,16 @@
 // in which the object is found
 void to_json(nlohmann::json & json, const MooseApp & app); // MooseDocs:to_json
+
+namespace libMesh
+{
+void to_json(nlohmann::json & json, const Point & p);
+}
```
```suggestion
void to_json(nlohmann::json & json, const libMesh::Point & p);
```

GiudGiud replied:
I think this is needed actually.
See comment above, needs to be instantiated in namespace it's used

lindsayad replied:
Alright. Can you add to the comment and say 
```
// Specializations for to_json, which _must_ be specialized in the namespace                                          
// in which the object is found in order to enable argument-dependent lookup
```

lindsayad replied:
And it would also be helpful to add a comment in `ReporterContext::store` about how `nlohmann` qualification is needed because `std::vector` specialization is *not* in the `std` namespace, it's in the `nlohmann` namespace. (I know this is not your work, but this was pretty confusing as i was trying to understand why my suggested change was not working). I could also do this in a separate PR if you would prefer.

loganharbour replied:
For your first comment, this already exists: https://github.com/idaholab/moose/blob/2a048be44daac9b4c7fb3f3c265284df72987408/framework/include/outputs/JsonIO.h#LL16C32-L16C32

> // Specializations for to_json, which _must_ be specialized in the namespace
> // in which the object is found

lindsayad replied:
Yea it needs more meaning. Just saying it has to be done is not enough for me. It's nice to know why

lindsayad replied:
You'll notice that I built my addition off of your comment

lindsayad replied:
So in full I would have the comment be
```
// Overloads for to_json, which _must_ be overloaded in the namespace                                          
// in which the object is found in order to enable argument-dependent lookup
// See https://en.cppreference.com/w/cpp/language/adl for more information
```
I've also clarified the language to "overload" instead of "specialization" since these are not specializations in the C++ terminology

loganharbour replied:
Sounds good, just wanted to make sure there wasn't a duplicate

lindsayad commented on framework/include/positions/ElementCentroidPositions.h:
```diff
@@ -0,0 +1,29 @@
+  ElementCentroidPositions(const InputParameters & parameters);
+  virtual ~ElementCentroidPositions() = default;
+
+  void initialize() override;
```
```suggestion
  virtual void initialize() override;
```

GiudGiud committed:
First round of review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

lindsayad commented on framework/src/base/Moose.C:
```diff
@@ -309,6 +310,7 @@ addActionTypes(Syntax & syntax)
                            "(setup_dampers)"
                            "(setup_residual_debug)"
                            "(add_bounds_vectors)"
+                           "(add_positions)"
                            "(add_multi_app)"
```
I assume there was a reason for this ordering. I saw at least one place (probably `MultiAppPositions`) where you had to delay data construction until `initialize` because of this ordering ... but I'm guessing the dependency issues would be worse if you flipped these

lindsayad commented on framework/src/dirackernels/ReporterPointSource.C:
```diff
@@ -19,57 +19,97 @@ ReporterPointSource::validParams()
 {
+  if (isParamValid("point_name") == (isParamValid("x_coord_name") && isParamValid("y_coord_name") &&
+                                     isParamValid("z_coord_name")))
+  {
+    paramError("Either supply x,y, and z reporters or a point reporter.");
+  }
```
```suggestion
    paramError("Either supply x,y, and z reporters or a point reporter.");
```

lindsayad commented on framework/src/dirackernels/ReporterPointSource.C:
```diff
@@ -19,57 +19,97 @@ ReporterPointSource::validParams()
+    if (_values.size() != _x_coord.size() || _values.size() != _y_coord.size() ||
+        _values.size() != _z_coord.size())
+    {
+      std::string errMsg =
```
```suggestion
      const std::string errMsg =
```

lindsayad commented on framework/src/dirackernels/ReporterPointSource.C:
```diff
@@ -19,57 +19,97 @@ ReporterPointSource::validParams()
+    _point_to_index.clear();
+    for (std::size_t i = 0; i < _values.size(); ++i)
+    {
+      Point pt(_x_coord[i], _y_coord[i], _z_coord[i]);
```
```suggestion
      const Point pt(_x_coord[i], _y_coord[i], _z_coord[i]);
```

lindsayad commented on framework/src/dirackernels/ReporterPointSource.C:
```diff
@@ -19,57 +19,97 @@ ReporterPointSource::validParams()
+
+      mooseError(errMsg);
+    }
+    for (MooseIndex(_values) i = 0; i < _values.size(); ++i)
```
```suggestion
    for (const auto i : index_range(_values))
```

lindsayad commented on framework/src/multiapps/CentroidMultiApp.C:
```diff
@@ -16,6 +16,7 @@
 #include "libmesh/parallel_algebra.h"
 
 registerMooseObject("MooseApp", CentroidMultiApp);
+// TODO: Deprecate and use Positions system
```
```suggestion

// TODO: Deprecate and use Positions system
```

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -517,6 +532,52 @@ MultiApp::fillPositions()
   }
+  else if (isParamValid("positions_objects"))
+  {
+    std::vector<PositionsName> positions_param_objs =
+        getParam<std::vector<PositionsName>>("positions_objects");
+    std::vector<FileName> input_files = getParam<std::vector<FileName>>("input_files");
```
```suggestion
    const auto & positions_param_objs =
        getParam<std::vector<PositionsName>>("positions_objects");
    const auto & input_files = getParam<std::vector<FileName>>("input_files");
```

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -517,6 +532,52 @@ MultiApp::fillPositions()
+    // Keeps track of where each positions object start in terms of subapp numbers
+    unsigned int offset = 0;
+
+    for (unsigned int p_obj_it = 0; p_obj_it < positions_param_objs.size(); p_obj_it++)
```
```suggestion
    for (const auto p_obj_it : index_range(positions_param_objs))
```

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -517,6 +532,52 @@ MultiApp::fillPositions()
+
+    for (unsigned int p_obj_it = 0; p_obj_it < positions_param_objs.size(); p_obj_it++)
+    {
+      const std::string positions_name = positions_param_objs[p_obj_it];
```
```suggestion
      const std::string & positions_name = positions_param_objs[p_obj_it];
```

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -517,6 +532,52 @@ MultiApp::fillPositions()
+      const std::string positions_name = positions_param_objs[p_obj_it];
+      auto positions_obj = &_fe_problem.getPositionsObject(positions_name);
+
+      auto data = positions_obj->getPositions(true);
```
```suggestion
      const auto & data = positions_obj->getPositions(true);
```

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -517,6 +532,52 @@ MultiApp::fillPositions()
+
+      auto data = positions_obj->getPositions(true);
+
+      for (const auto & d : data)
+        _positions.push_back(d);
```
```suggestion
      _positions = data;
```

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -517,6 +532,52 @@ MultiApp::fillPositions()
+                   "'" + positions_name + "' is not of the expected type. Should be a Positions");
+
+      // Keep track of which positions is tied to what subapp
+      for (unsigned int i = 0; i < data.size(); ++i)
```
```suggestion
      for (const auto i : index_range(data))
```

GiudGiud replied:
keeping old one.
with new one, i is reported as unused and would need some trick to bypass that

GiudGiud commented on framework/include/positions/MultiAppPositions.h:
```diff
@@ -0,0 +1,31 @@
+  void initialize() override;
+  using Positions::_positions;
+
+  /// Whether to use the subapp mesh centroids to compute the positions, in addition to the positions
```
```suggestion
  /// Whether to use the subapp mesh centroids to compute the positions, further translated by the positions
```

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -1234,3 +1301,14 @@ MultiApp::getMultiAppName(const std::string & base_name, dof_id_type index, dof_
+MultiApp::position(unsigned int app) const
+{
+  // If we're not using positions, it won't have changed
+  if (!_positions_objs.size())
```
```suggestion
  if (_positions_objs.empty())
```

GiudGiud committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

lindsayad commented on framework/src/outputs/JsonIO.C:
```diff
@@ -40,3 +40,14 @@ to_json(nlohmann::json & json, const MooseApp & app)
 }
 // MooseDocs:to_json_end
+
+namespace libMesh
+{
+void
+to_json(nlohmann::json & json, const Point & p)
+{
+  json["x"] = p(0);
+  json["y"] = p(1);
+  json["z"] = p(2);
+}
+}
```
```suggestion
void
to_json(nlohmann::json & json, const libMesh::Point & p)
{
  json["x"] = p(0);
  json["y"] = p(1);
  json["z"] = p(2);
}
```

lindsayad commented on framework/src/positions/ElementCentroidPositions.C:
```diff
@@ -0,0 +1,67 @@
+  }
+  else
+  {
+    _positions.resize(_mesh.getMesh().n_active_local_elem());
```
Are these not supposed to be parallel consistent?

GiudGiud replied:
the reporter system lets you declare how you are handling replication OR distribution

but it doesnt do it for you.
we set it to the size you see for this domain, then base class gathers

GiudGiud commented on framework/include/positions/Positions.h:
```diff
@@ -0,0 +1,92 @@
+  ///{
+  /// Getters for the positions vector for the desired dimension
+  /// 1D will be the only one guaranteed to succeed
+  const std::vector<Point> & getPositions(bool initial = false) const;
```
we can, removing them

lindsayad commented on framework/src/positions/ElementGroupCentroidPositions.C:
```diff
@@ -0,0 +1,351 @@
+{
+  InputParameters params = Positions::validParams();
+  params += BlockRestrictable::validParams();
+  params.addClassDescription("Gets the Positions of the centroid of groups elements. "
```
```suggestion
  params.addClassDescription("Gets the Positions of the centroid of groups of elements. "
```

lindsayad commented on framework/src/positions/ElementGroupCentroidPositions.C:
```diff
@@ -0,0 +1,351 @@
+  _mesh.errorIfDistributedMesh(type());
+
+  // We are not excluding using both block restriction and extra element ids
+  if (_group_type == "extra_id" || _group_type == "block_and_ids")
```
```suggestion
  if (_group_type == "extra_id" || _group_type == "block_and_extra_id")
```

GiudGiud commented on framework/include/positions/Positions.h:
```diff
@@ -0,0 +1,92 @@
+  const bool _need_sort;
+};
+
+void dataStore(std::ostream & stream, const std::vector<Point> & positions, void * context);
+void dataLoad(std::istream & stream, const std::vector<Point> & positions, void * context);
```
seeing compilation errors without this. I think we need to instantiate the template explcitity in the header. Having it in the source is not enough

lindsayad replied:
I just deleted these lines and didn't see any compile errors when compiling the framework

GiudGiud replied:
i ll try again. maybe I fooled myself

lindsayad commented on framework/src/positions/ElementGroupCentroidPositions.C:
```diff
@@ -0,0 +1,351 @@
+
+  // If users did not specify a value for an extra element integer, they want all the bins
+  // We'll need to loop through the mesh to find the element extra ids
+  for (MooseIndex(_extra_id_group_indices) i = 0; i < _extra_id_group_indices.size(); i++)
```
```suggestion
  for (const auto i : index_range(_extra_id_group_indices))
```

lindsayad commented on framework/src/positions/ElementGroupCentroidPositions.C:
```diff
@@ -0,0 +1,351 @@
+  for (MooseIndex(_extra_id_group_indices) i = 0; i < _extra_id_group_indices.size(); i++)
+  {
+    auto & indices = _extra_id_group_indices[i];
+    if (indices.size() == 0)
```
```suggestion
    if (indices.empty())
```

lindsayad commented on framework/src/positions/ElementGroupCentroidPositions.C:
```diff
@@ -0,0 +1,351 @@
+      _extra_id_group_indices.size());
+
+  // Make index maps to go from the extra element id to the positions array index
+  for (MooseIndex(_extra_id_group_indices) i = 0; i < _extra_id_group_indices.size(); i++)
```
```suggestion
  for (const auto i : index_range(_extra_id_group_indices))
```

lindsayad commented on framework/src/positions/ElementGroupCentroidPositions.C:
```diff
@@ -0,0 +1,351 @@
+    // Keeps track of indices in multi-D arrays
+    std::vector<unsigned int> previous_indices(4);
+
+    for (MooseIndex(_extra_id_names) i = 0; i < _extra_id_names.size(); i++)
```
```suggestion
    for (const auto i : index_range(_extra_id_names))
```

lindsayad commented on framework/src/positions/ElementGroupCentroidPositions.C:
```diff
@@ -0,0 +1,351 @@
+
+  // Renormalize by the total bin volumes
+  if (_extra_id_names.size() == 1)
+    for (MooseIndex(_positions) i = 0; i < _positions.size(); i++)
```
I would replace all uses of `MooseIndex` with `index_range`

GiudGiud replied:
will do

GiudGiud replied:
Does index_range work with a delete nested?
I get crashes

lindsayad replied:
Oh yea erasing invalidates the end iterator so can't use the `index_range` in that case

lindsayad commented on framework/src/positions/FilePositions.C:
```diff
@@ -0,0 +1,51 @@
+
+FilePositions::FilePositions(const InputParameters & parameters) : Positions(parameters)
+{
+  std::vector<FileName> positions_files = getParam<std::vector<FileName>>("files");
```
```suggestion
  const auto & positions_files = getParam<std::vector<FileName>>("files");
```

lindsayad commented on framework/src/positions/FilePositions.C:
```diff
@@ -0,0 +1,51 @@
+  _positions_2d.resize(positions_files.size());
+
+  // Copied from MultiApp.C
+  for (unsigned int p_file_it = 0; p_file_it < positions_files.size(); p_file_it++)
```
```suggestion
  for (const auto p_file_it : index_range(positions_files))
```

lindsayad commented on framework/src/positions/MultiAppPositions.C:
```diff
@@ -0,0 +1,91 @@
+{
+  clearPositions();
+
+  std::vector<MultiAppName> multiapps = getParam<std::vector<MultiAppName>>("multiapps");
```
```suggestion
  const auto & multiapps = getParam<std::vector<MultiAppName>>("multiapps");
```

lindsayad commented on framework/src/positions/MultiAppPositions.C:
```diff
@@ -0,0 +1,91 @@
+  std::vector<MultiAppName> multiapps = getParam<std::vector<MultiAppName>>("multiapps");
+  _positions_2d.resize(multiapps.size());
+
+  for (unsigned int m_it = 0; m_it < multiapps.size(); m_it++)
```
```suggestion
  for (const auto m_it : index_range(multiapps))
```

lindsayad commented on framework/src/positions/MultiAppPositions.C:
```diff
@@ -0,0 +1,91 @@
+
+  for (unsigned int m_it = 0; m_it < multiapps.size(); m_it++)
+  {
+    const std::string multiapp_name = multiapps[m_it];
```
```suggestion
    const std::string & multiapp_name = multiapps[m_it];
```

lindsayad commented on framework/src/positions/MultiAppPositions.C:
```diff
@@ -0,0 +1,91 @@
+
+    for (const auto & i_global : make_range(multiapp->numGlobalApps()))
+    {
+      auto p = multiapp->position(i_global);
```
```suggestion
      const auto p = multiapp->position(i_global);
```

lindsayad commented on framework/src/positions/MultiAppPositions.C:
```diff
@@ -0,0 +1,91 @@
+                                 fe_problem_base.getDisplacedProblem().get() != NULL)
+                                    ? fe_problem_base.getDisplacedProblem()->mesh().getMesh()
+                                    : fe_problem_base.mesh().getMesh();
+        Point centroid = MooseMeshUtils::meshCentroidCalculator(mesh);
```
```suggestion
        const Point centroid = MooseMeshUtils::meshCentroidCalculator(mesh);
```

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -0,0 +1,244 @@
+{
+  // TODO Use faster & fancier machinery such as a KNN-partition
+  unsigned int nearest_index = 0;
+  Real nearest_distance_sq = std::numeric_limits<float>::max();
```
```suggestion
  auto nearest_distance_sq = std::numeric_limits<Real>::max();
```

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -0,0 +1,244 @@
+Positions::getNearestPosition(const Point & target) const
+{
+  // TODO Use faster & fancier machinery such as a KNN-partition
+  unsigned int nearest_index = 0;
```
```suggestion
  std::size_t nearest_index = 0;
```

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -0,0 +1,244 @@
+  // TODO Use faster & fancier machinery such as a KNN-partition
+  unsigned int nearest_index = 0;
+  Real nearest_distance_sq = std::numeric_limits<float>::max();
+  for (const auto & pt : _positions)
```
```suggestion
  for (const auto i : index_range(_positions))
  {
    const auto & pt = _positions[i];
```

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -0,0 +1,244 @@
+  for (const auto & pt : _positions)
+    if ((pt - target).norm_sq() < nearest_distance_sq)
+    {
+      nearest_index = &pt - &_positions[0];
```
```suggestion
      nearest_index = i;
```

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -0,0 +1,244 @@
+    {
+      nearest_index = &pt - &_positions[0];
+      nearest_distance_sq = (pt - target).norm_sq();
+    }
```
```suggestion
    }
  }
```

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -0,0 +1,244 @@
+  std::vector<Point> temp_4d_unrolled;
+  for (auto vec : _positions_2d)
+    temp_2d_unrolled.insert(temp_2d_unrolled.end(), vec.begin(), vec.end());
+  for (auto vec_vec : _positions_3d)
+    for (auto vec : vec_vec)
+      temp_3d_unrolled.insert(temp_3d_unrolled.end(), vec.begin(), vec.end());
+  for (auto vec_vec_vec : _positions_4d)
+    for (auto vec_vec : vec_vec_vec)
+      for (auto vec : vec_vec)
```
```suggestion
  for (const auto & vec_vec : _positions_3d)
    for (const auto & vec : vec_vec)
      temp_3d_unrolled.insert(temp_3d_unrolled.end(), vec.begin(), vec.end());
  for (const auto & vec_vec_vec : _positions_4d)
    for (const auto & vec_vec : vec_vec_vec)
      for (const auto & vec : vec_vec)
```

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -0,0 +1,244 @@
+
+    // Gather all components of all positions
+    std::vector<Real> x(_positions.size()), y(_positions.size()), z(_positions.size());
+    for (const auto & pt : _positions)
+    {
+      x[&pt - &_positions[0]] = pt(0);
+      y[&pt - &_positions[0]] = pt(1);
+      z[&pt - &_positions[0]] = pt(2);
+    }
```
I think the idiom where you loop over indexes is clearer to read

GiudGiud replied:
will change if cant use parallel algebra

GiudGiud replied:
Removed those loops

lindsayad commented on framework/src/positions/Positions.C:
```diff
@@ -0,0 +1,244 @@
+
+  if (_need_broadcast)
+  {
+    // TODO: be able to do this
+    // _communicator.allgather(_positions, /* identical buffer lengths = */ false);
```
Why can't we do this? If you include `libmesh/parallel_algebra.h` I feel like this should work

GiudGiud replied:
indeed

lindsayad commented on framework/src/positions/ReporterPositions.C:
```diff
@@ -0,0 +1,64 @@
+{
+  clearPositions();
+
+  std::vector<ReporterName> positions_reporters = getParam<std::vector<ReporterName>>("reporters");
```
```suggestion
  const auto & positions_reporters = getParam<std::vector<ReporterName>>("reporters");
```

lindsayad commented on framework/src/positions/ReporterPositions.C:
```diff
@@ -0,0 +1,64 @@
+  std::vector<ReporterName> positions_reporters = getParam<std::vector<ReporterName>>("reporters");
+  _positions_2d.resize(positions_reporters.size());
+
+  for (unsigned int r_it = 0; r_it < positions_reporters.size(); r_it++)
+  {
+    const std::string reporter_name = positions_reporters[r_it];
```
```suggestion
  for (const auto r_it : index_range(positions_reporters))
  {
    const std::string & reporter_name = positions_reporters[r_it];
```

lindsayad commented on test/tests/positions/creating_multiapps/tests:
```diff
@@ -0,0 +1,49 @@
+    input = 'apps_from_positions.i'
+    jsondiff = 'apps_from_positions_out.json'
+
+    requirement = "The system shall be able to create multiapps from Positions objects."
```
```suggestion
    requirement = "The system shall be able to create multiapps from positions objects."
```

lindsayad commented on test/tests/positions/creating_multiapps/tests:
```diff
@@ -0,0 +1,49 @@
+    # - the multiapps are created using the initial positions, however, when
+    # the positions are retrieved for output, the positions have been updated
+
+    requirement = "The system shall be able to create multiapps from Positions objects using their initial positions."
```
```suggestion
    requirement = "The system shall be able to create multiapps from positions objects using their initial positions."
```

lindsayad commented on test/tests/positions/creating_multiapps/tests:
```diff
@@ -0,0 +1,49 @@
+      type = RunException
+      input = 'apps_from_positions.i'
+      cli_args = 'MultiApps/m1/positions_objects=results_m1'
+      detail = 'the positions objects to use for creating multiapps are not initialized,'
```
```suggestion
      detail = 'the positions objects to use for creating multiapps are not initialized, or'
```

lindsayad commented on test/tests/positions/reporter_positions.i:
```diff
@@ -0,0 +1,42 @@
+[]
+
+[Positions]
+  [input]
+    type = InputPositions
+    positions = '0 0 1
+                 1 0 2'
+    outputs = none
+  []
+  [input_2]
+    type = InputPositions
+    positions = '0.1 0.1 1
+                 1.2 0 2.2'
+    outputs = none
+  []
+  [reporter]
+    type = ReporterPositions
+    reporters = 'input/positions_1d input_2/positions_1d'
+  []
```
Dumb question. I thought there was only ever one `Positions` object based on the assertion in `FEProblemBase`?

lindsayad replied:
Oh never mind. It was just one per name

lindsayad reviewed (changes requested):
Alright made it through

GiudGiud commented on framework/include/positions/ElementGroupCentroidPositions.h:
```diff
@@ -0,0 +1,48 @@
+  virtual ~ElementGroupCentroidPositions() = default;
+
+  void initialize() override;
+  using Positions::_positions;
```
indeed. Cant remember why i added it. It doesnt seem necessary anymore

GiudGiud commented on test/tests/positions/element_group_centroid_positions.i:
```diff
@@ -0,0 +1,65 @@
+    type = ElementGroupCentroidPositions
+    extra_id_name = 'assembly_id pin_id material_id'
+    extra_id = '0; 1 2 4 6; ;'
+    grouping_type = 'block_and_ids'
```
```suggestion
    grouping_type = 'block_and_extra_id'
```

GiudGiud commented on test/tests/positions/element_group_centroid_positions.i:
```diff
@@ -0,0 +1,65 @@
+    block = '1 2'
+    extra_id_name = 'assembly_id pin_id material_id'
+    extra_id = '0; 1 2 4 6; ;'
+    grouping_type = 'block_and_ids'
```
```suggestion
    grouping_type = 'block_and_extra_id'
```

GiudGiud commented on test/tests/positions/element_group_centroid_positions.i:
```diff
@@ -0,0 +1,65 @@
+    block = '1 2'
+    extra_id_name = 'pin_id'
+    extra_id = '1 2 4'
+    grouping_type = 'block_and_ids'
```
```suggestion
    grouping_type = 'block_and_extra_id'
```

GiudGiud commented on test/tests/positions/element_group_centroid_positions.i:
```diff
@@ -0,0 +1,65 @@
+    block = '1 2'
+    extra_id_name = 'assembly_id pin_id'
+    extra_id = '1; 1 2 4'
+    grouping_type = 'block_and_ids'
```
```suggestion
    grouping_type = 'block_and_extra_id'
```

GiudGiud committed:
Apply suggestions from code review
- more constness and more references
- use index-range when possible
- some comments to help with understanding ADL issues

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

lindsayad commented on framework/src/outputs/JsonIO.C:
```diff
@@ -44,7 +44,7 @@ to_json(nlohmann::json & json, const MooseApp & app)
 {
 void
-to_json(nlohmann::json & json, const Point & p)
+to_json(nlohmann::json & json, const libMesh::Point & p)
```
With the declaration and definition now in the `libMesh` namespace this isn't necessary anymore. (But it's also not a problem either)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1502463/) on 1f26c28 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24071/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24071/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2a048be44daac9b4c7fb3f3c265284df72987408`

lindsayad commented on framework/include/reporters/ReporterContext.h:
```diff
@@ -85,6 +85,9 @@ class ReporterContextBase : public libMesh::ParallelObject
   /// and include all the data including the old values.
   ///
   /// This method only outputs the current value within the JSONOutput object.
+  /// NOTE: nlohmann qualification is needed for argument json because std::vector specialization is
```
I said "specialization" earlier but pedantically it's "overload"

GiudGiud replied:
can I fix it in next PR? Follow up for capability w/ transfers

lindsayad commented:
I'll approve this when the framework targets are green and coverage report looks good

moosebuild commented:
All [jobs](https://civet.inl.gov/event/122446/) on 082ec4c : invalidated by @lindsayad

lindsayad commented on framework/src/positions/Positions.C:
There's quite a bit of coverage missing in this file. Will that be addressed in the next PR?

GiudGiud replied:
Positions::getNearestPosition(const Point & target) const
will be used for nearest-node transfers

The multi-D getters might only get used if @zachmprince changes his mind.
The rest are exceptions which I could try to hit in the next PR or here if you want me to

lindsayad replied:
I don't care about the exception testing

lindsayad reviewed (approved):
Very cool development!

GiudGiud commented:
Test failure looks totally unrelated but it also wont recover from invalidation.

lindsayad reviewed (changes requested):
I guess I would be in big trouble per today's meeting if I didn't request a newsletter entry. I'm also leaning towards adding a commit that removes those multi-D functions since it sounds like they may never be used. If we want them again, we could cleanly revert that commit. Up to you tho

GiudGiud committed:
Add one more tex garbage output to gitignore

GiudGiud committed:
Add unit test for Positions, refs #23587

GiudGiud committed:
Add newsletter entry for Positions, refs #23587

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1512019/) on 0edfb7a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24071/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24071/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d2987036293b03cc8fe4fcc45c7a9daa3db65fac`

GiudGiud commented:
Added a unit test, newsletter and rebased on next for the conflict with April's change

I d like to keep the multiD capability. This stuff clearly has to be ordered. I would like to see 1 Positions object for the reactor, not 1 for each of the 193 or so assemblies, and a grouping by assembly makes sense for pins for example

GiudGiud commented:
Thanks for the reviews!
